File: aws-ec2/operation/add_security_group_rule.rb
=begin
##########################################################################
#
#  INFIVERVE TECHNOLOGIES PTE LIMITED CONFIDENTIAL
#  __________________
# 
#  (C) INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE
#  All Rights Reserved.
#  Product / Project: Flint IT Automation Platform
#  NOTICE:  All information contained herein is, and remains
#  the property of INFIVERVE TECHNOLOGIES PTE LIMITED.
#  The intellectual and technical concepts contained
#  herein are proprietary to INFIVERVE TECHNOLOGIES PTE LIMITED.
#  Dissemination of this information or any form of reproduction of this material
#  is strictly forbidden unless prior written permission is obtained
#  from INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE.
=end

@log.trace("Started executing 'fb-cloud:aws-ec2:operation:add_security_group_rule.rb' flintbit...")
begin
connector_name = @input.get("connector_name")
direction = @input.get("direction")
from_port = @input.get("from_port")
to_port = @input.get("to_port")
protocol = @input.get("protocol")
group_id = @input.get("security_group_id")

# Optional
ip_range = @input.get("ip_range")
cidr_block = @input.get("cidr_block")
region = @input.get("region")

request_timeout = @input.get('timeout')
@log.trace("Calling AWS-EC2 Connector...")
@log.info("Flintbit input parameters are connector_name : #{connector_name} | direction : #{direction}
| from_port :#{from_port} | to_port : #{to_port} | protocol : #{protocol} | group_id : #{group_id}
 | access_key : #{@access_key} | secret_key : #{@secret_key}")

connector_call = @call.connector(connector_name)
                .set("action","security-group-add-rule")
                .set("access-key",@input.get("access-key"))
                .set("security-key",@input.get("security-key"))

                    if !group_id.nil? && !group_id.empty?
                      connector_call.set("security-group-id",group_id)
                    else
                      raise "Please provide 'Security Group Id'"
                    end
                    if !direction.nil? && !direction.empty?
                      connector_call.set("direction",direction)
                    else
                      raise "Please provide 'direction' for security group rule"
                    end
                    if !from_port.nil? && !from_port.empty?
                      connector_call.set("from-port",from_port.to_i)
                    else
                      raise "Please provide 'from port' for security group rule"
                    end
                    if !to_port.nil? && !to_port.empty?
                      connector_call.set("to-port",to_port.to_i)
                    else
                      raise "Please provide 'to port' for security group rule"
                    end
                    if  !protocol.nil? && !protocol.empty?
                      connector_call.set("protocol",protocol)
                    else
                      raise "Please provide 'protocol' for security group rule"
                    end
                    if !request_timeout.nil? && !request_timeout.empty?
                      connector_call.timeout(request_timeout)
                    else
                      @log.trace("Calling #{connector_name} with default timeout...")
                    end
                    if !region.nil? && !region.empty?
                           connector_call.set('region', region)
                    else
                          @log.trace("region is not provided so using default region 'us-east-1'")
                    end
                    if  !cidr_block.nil? &&  !cidr_block.empty?
                      if  !ip_range.nil? && !ip_range.empty?
                        raise "Please provide either CIDR block or IP ranges for security group rule. Do not provide both at same time."
                      else
                      response = connector_call.set("cidr-block",cidr_block).sync
                      end
                    elsif  !ip_range.nil? && !ip_range.empty?
                      if  !cidr_block.nil? &&  !cidr_block.empty?
                        raise "Please provide either CIDR block or IP ranges for security group rule. Do not provide both at same time."
                      else
                      response = connector_call.set("ip-range",ip_range).sync
                      end
                    else
                      raise "Please provide either CIDR block or IP ranges for security group rule"
                    end

  #Amazon EC2 Connector Response Meta Parameters
   response_exitcode = response.exitcode #Exit status code
   response_message = response.message   #Execution status messages

  if response_exitcode == 0
     @log.info("Response ::#{response}")
     @log.info("Success in executing AWS-EC2 Connector where, exitcode :: #{response_exitcode}| message :: #{response_message}")
     @user_message = "Successfully added the security rules."
     @output.set("exit-code",response_exitcode).set("message",response_message).set("user_message",@user_message)
  else
     @log.error("Response ::#{response}")
     @log.error("Failure in executing AWS-EC2 Connector where, exitcode :: #{response_exitcode} | message :: #{response_message}")
     @user_message = "Failed to add the security rules."
     response=response.to_s
     if !response.empty?
     @output.set('message', response_message).set('exit-code', 1).setraw('error-details',response.to_s).set("user_message",@user_message)
     else
     @output.set('message', response_message).set('exit-code', 1).set("user_message",@user_message)
     end
  end



rescue => e
    @log.error(e.message)
    @output.set('message', e.message).set('exit-code', -1)
end
@log.trace("Finished execution of 'fb-cloud:aws-ec2:operation:add_security_group_rule.rb' flintbit...")

File: aws-ec2/operation/workflow/add_security_group_rule.js
/*************************************************************************
 * 
 * INFIVERVE TECHNOLOGIES PTE LIMITED CONFIDENTIAL
 * __________________
 * 
 * (C) INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE
 * All Rights Reserved.
 * Product / Project: Flint IT Automation Platform
 * NOTICE:  All information contained herein is, and remains
 * the property of INFIVERVE TECHNOLOGIES PTE LIMITED.
 * The intellectual and technical concepts contained
 * herein are proprietary to INFIVERVE TECHNOLOGIES PTE LIMITED.
 * Dissemination of this information or any form of reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE.
 */

log.trace("Started executing the flintbit 'fb-cloud:aws-ec2:operation:workflow:add_security_group_rule.js'");

input_scope = JSON.parse(input);   //For checking if the input contains the necessary keys

//Connector Name  - mandatory
connector_name = "amazon-ec2";
connector_call = call.connector(connector_name);
log.info("Connector Name: "+connector_name);

//Action - mandatory
action = "security-group-add-rule";
log.info("Action: "+action);

if(input_scope.hasOwnProperty("cloud_connection")){

    //Access Key - mandatory
    access_key = input_scope.cloud_connection.encryptedCredentials["access_key"];
    if(access_key!=null || access_key!=""){
       connector_call.set("access-key",access_key); 
       log.info("Access Key is given");
    }
    else{
        log.error("Access-Key is null or empty string.");
    }

    //Security Key - mandatory
    secret_key = input_scope.cloud_connection.encryptedCredentials["secret_key"];
    if(secret_key!=null || secret_key!=""){
       connector_call.set("security-key",secret_key); 
       log.info("Security Key is given");
    }
    else{
        log.error("Security-Key is null or empty string.");
    }

    //Security Group ID - mandatory
    if(input_scope.hasOwnProperty("security_group_id")){
        security_group_id = input.get("security_group_id");
        if(security_group_id!=null || security_group_id!=""){
            connector_call.set("security-group-id",security_group_id); 
            log.info("Security Group ID: "+security_group_id);
        }
        else{
            log.error("Security Group ID is null or empty string.")
        }
    }
    else{  //security_group_id key not present in input JSON 
        log.error("Input does not contain the key 'security_group_id'")
    }

    //Direction - mandatory
    if(input_scope.hasOwnProperty("direction")){
        direction = input.get("direction");
        if(direction!=null || direction!=""){
            connector_call.set("direction",direction); 
            log.info("Direction: "+direction);
        }
        else{
            log.error("Direction is null or empty string.")
        }
    }
    else{  //direction key not present in input JSON 
        log.error("Input does not contain the key 'direction'")
    }

    //Protocol - mandatory
    if(input_scope.hasOwnProperty("protocol")){
        protocol = input.get("protocol");
        if(protocol!=null || protocol!=""){
            connector_call.set("protocol",protocol); 
            log.info("Protocol: "+protocol);
        }
        else{
            log.error("Protocol is null or empty string.")
        }
    }
    else{  //protocol key not present in input JSON 
        log.error("Input does not contain the key 'protocol'")
    }

    //From port - mandatory
    if(input_scope.hasOwnProperty("from_port")){
        from_port = input.get("from_port");
        if(from_port!=null || from_port!=""){
            connector_call.set("from-port",from_port); 
            log.info("From port: "+from_port);
        }
        else{
            log.error("from_port is null or empty string.")
        }
    }
    else{  //from_port key not present in input JSON 
        log.error("Input does not contain the key 'from_port'")
    }

    //To port - mandatory
    if(input_scope.hasOwnProperty("to_port")){
        to_port = input.get("to_port");
        if(to_port!=null || to_port!=""){
            connector_call.set("to-port",to_port); 
            log.info("To port: "+to_port);
        }
        else{
            log.error("to_port is null or empty string.")
        }
    }
    else{  //to_port key not present in input JSON 
        log.error("Input does not contain the key 'to_port'")
    }

    //Region - mandatory
    if(input_scope.hasOwnProperty("region")){
        region = input.get("region");
        if(region!=null || region!=""){
            connector_call.set("region",region); 
            log.info("Region: "+region);
        }
        else{
            log.error("Region is null or empty string.");
        }
    }
    else{  //region key not present in input JSON 
        log.error("Input does not contain the key 'region'");
    }

    //IP Range - NOT mandatory
    if(input_scope.hasOwnProperty("ip_range")){
        ip_range = input.get("ip_range");
        if(ip_range!=null || ip_range!=""){
            connector_call.set("ip-range",ip_range); 
            log.info("IP range: "+ip_range);
        }
        else{
            log.info("IP Range not given");
        }
    }
    else{
        log.info("Input does not contain the key 'ip_range'");
    }

    //CIDR Block - NOT mandatory
    if(input_scope.hasOwnProperty("cidr_block")){
        cidr_block = input.get("cidr_block");
        if(cidr_block!=null || cidr_block!=""){
            connector_call.set("cidr-block",cidr_block); 
            log.info("CIDR block: "+cidr_block);
        }
        else{
            log.info("CIDR Block not given");
        }
    }
    else{
        log.info("Input does not contain the key 'cidr_block'");
    }

    //Timeout - NOT mandatory
    if(input_scope.hasOwnProperty("request_timeout")){
        request_timeout = input.get("request_timeout");
        if(request_timeout!=null || request_timeout!=""){
            connector_call.set("timeout",request_timeout); 
            log.info("Request Timeout: "+request_timeout);
        }
        else{
            connector_call.set("timeout",240000); 
            log.info("request_timeout not given. Setting 240000 miliseconds as timeout");
        }
    }
    else{
        connector_call.set("timeout",240000); 
        log.info("request_timeout not given. Setting 240000 miliseconds as timeout");
    }

    //Connector call
    response = connector_call.set("action",action).sync();

    //Response meta-parameters
    response_exitcode = response.exitcode();
    response_message = response.message();

    if(response_exitcode==0){
        log.info("Successfully added a Security Group Rule");
        output.set("message",response_message)
            .set("exit-code",response_exitcode)
            .set("user_message","Successfully added a Security Group Rule");
        log.trace("Finished executing 'fb-cloud:aws-ec2:operation:workflow:add_security_group_rule.js' successfully");
    }
    else{
        log.error("Failure in execution, message:"+response_message+" | exitcode:"+response_exitcode);
        output.set("error",response_message).set("exit-code",-1);
        log.trace("finished executing 'fb-cloud:aws-ec2:operation:workflow:add_security_group_rule.js' with errors")
    }
}
else{ //cloud_connection key not present in input JSON
    log.error("Cloud Connection not given. Can not authenticate without Secret-Key and Access-Key");
}
